probe-rs/src/probe/jlink/mod.rs

```diff
@@ -447,7 +533,263 @@ impl JTAGAccess for JLink {
+            .chain(iter::repeat(false).take(1)) // Receive 1 Parity bit.
+            .chain(iter::repeat(false).take(1)); // Receive 1 Turnaround bit.
+
+        // Now we try to issue the request until it fails or succeeds.
+        // If we timeout we retry a maximum of 5 times.
+        let mut retries = 0;
+        while retries < 5 {
+            // Transmit the sequence and record the line sequence for the ack bits.
+            let mut result_sequence = self
+                .handle
+                .get_mut()
+                .unwrap()
+                .swd_io(direction.clone(), swd_io_sequence.iter().copied())?;
+
+            // Throw away the two idle bits.
+            result_sequence.split_off(2);
+            // Throw away the request bits.
+            result_sequence.split_off(8);
+
+            // Get the ack.
+            let ack = result_sequence.split_off(3).collect::<Vec<_>>();
+            if ack[1] {
+                // If ack[1] is set the host must retry the request. So let's do that right away!
+                retries += 1;
+                log::debug!("DAP line busy, retries remaining {}.", 5 - retries);
+                continue;
+            }
+            if ack[2] {
+                // A fault happened during operation.
+
+                // To get a clue about the actual fault we read the ctrl register,
+                // which will have the fault status flags set.
+                let response =
+                    DAPAccess::read_register(self, PortType::DebugPort, Ctrl::ADDRESS as u16)?;
+                let ctrl = Ctrl::from(response);
+                log::error!(
+                    "Reading DAP register failed. Ctrl/Stat register value is: {:#?}",
+                    ctrl
+                );
+
+                return Err(DebugProbeError::Unknown);
+            }
+
+            // If we are reading an AP register we only get the actual result in the next transaction.
+            // So we issue a special transaction to get the read value.
+            if port {
+                // We read the RDBUFF register to get the value of the last AP transaction.
+                // This special register just returns the last read value with no side-effects like auto-increment.
+                return DAPAccess::read_register(self, PortType::DebugPort, 0x0C);
+            } else {
+                // Take the data bits and convert them into a 32bit int.
+                let register_val = result_sequence.split_off(32);
+                let value = bits_to_byte(register_val);
+
+                // Make sure the parity is correct.
+                return if let Some(parity) = result_sequence.next() {
+                    if (value.count_ones() % 2 == 1) == parity {
+                        log::trace!("DAP read {}.", value);
+                        Ok(value)
+                    } else {
+                        log::error!("DAP read fault.");
+                        Err(DebugProbeError::Unknown)
```

Comment:
We should use a proper Error here, either a JLink specific one, or a more appropriate type of `DebugProbeError`.